identity-account-storage/src/types/key_location.rs

```diff
@@ -5,56 +5,60 @@ use core::fmt::Debug;
 use core::fmt::Display;
 use core::fmt::Formatter;
 use core::fmt::Result;
-use identity_core::common::Fragment;
+use identity_core::crypto::KeyType;
+use identity_did::verification::MethodData;
 use identity_did::verification::MethodType;
-use serde::Deserialize;
-use serde::Serialize;
-
-use crate::types::Generation;
+use identity_iota_core::document::IotaVerificationMethod;
+use rand::distributions::Alphanumeric;
+use rand::prelude::ThreadRng;
+use rand::Rng;
+use seahash::SeaHasher;
+use std::hash::Hash;
+use std::hash::Hasher;
 
 /// The storage location of a verification method key.
-#[derive(Clone, Hash, PartialEq, Eq, Deserialize, Serialize)]
+#[derive(Clone, Hash, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[non_exhaustive]
 pub struct KeyLocation {
-  method: MethodType,
-  fragment: Fragment,
-  generation: Generation,
+  pub key_type: KeyType,
+  pub fragment: String,
+  pub key_hash: u64,
 }
```

Comment:
We may want to keep/refactor `Fragment` since a `String` alone does not satisfy the constraints on DIDUrl fragments. Probably fine to use `String` for now though.

Reply:
Perhaps we can use `String` in the public interface for `KeyLocation`, but use `Fragment` internally, only to be able to consistently return the fragment without the leading `#`, independent of what the caller passed in? The `fragment` field would be made private, and a `KeyLocation::fragment` getter would be added, returning `Fragment::name`. `KeyLocation::fmt` would use that function then.

Reply:
That would be fine but I would add getters for all fields for consistency then.

As far as I see the `fragment` is only used in `remote_sign_data` internally, and that could be refactored to take a `VerificationMethod` directly (for the `DIDUrl`), bypassing this issue entirely so we can use `String` instead?

Reply:
I think you're right, but I'm not completely sure that we can always guarantee the higher-level validation, especially when we now expect storage implementations to construct key locations. Technically, of course, we will just pass them the fragment for them to just pass it into the constructor.
I've switched to `Fragment` now, but I'm also fine with changing it back. Not sure what to do.

Reply:
So the only reason we need it to be a valid DID-Url fragment right now is because the signature uses the `KeyLocation` fragment to reconstruct the DIDUrl of the `VerificationMethod`. Even then a `Fragment` just adds a `#` which is removed in the `KeyLocation::fragment()`, so it's not doing any validation anyway as far as I can see.

My initial comment referenced previous conversations/PRs where it was discussed that we should ideally refactor `Fragment` to do the same validation as `DIDUrl` if we intend to use it independently or remove it altogether.

I vote removing it and using `String` again for now, apologies for the back-and-forth.